### ISSUE TITLE
New weight initialisation when using dropout and ReLU

### ISSUE THREAD

**Issue by arnupretorius** [pytorch/pytorch#19536]
## 🚀 Feature
A new weight initialisation when using dropout with ReLU activation.

## Motivation
In a recent NeurIPS paper ([Pretorius et al. 2018](https://arxiv.org/pdf/1811.00293.pdf)), we derived a new weight initialisation strategy for deep ReLU networks that use dropout. This new initialisation ensures stable variance propagation for ReLU activations travelling through a dropout layer. The go-to initialisation for dense ReLU layers is the He initialisation, but in the above paper, we showed that if a dropout layer is applied after a dense ReLU layer the variances are longer preserved in the forward pass. This can lead to unstable signal propagation in deep ReLU networks that use dropout. We fix this with the new initialisation and thought it might be useful to the community if there was a PyTorch implementation.  

## Pitch

In short, the initialisation samples from a normal distribution with a `gain = \sqrt{2*(1-p)}`, where `p is the probability of an element to be zeroed`. To implement the initialisation in PyTorch, I was thinking perhaps something along these lines of code below could work? The first block of code could be added to the `calculate_gain` function. The second is then an implementation of the initialisation.

```python
elif nonlinearity == 'dropout_relu':
        if param is None:
            raise ValueError("dropout_rate set to {}. It is recommended to set the value equal to the dropout rate in the subsequent dropout layer".format(param))
        elif not isinstance(param, bool) and isinstance(param, int) or isinstance(param, float):
            # True/False are instances of int, hence check above
            if p < 0. or p > 1.:
                raise ValueError("dropout probability has to be between 0 and 1, "
                         "but got {}".format(p))
            else:
                dropout_rate = param
        else:
            raise ValueError("dropout_rate {} not a valid number".format(param))
        return math.sqrt(2.0 * (1-dropout_rate))
```

Below is the code implementing the dropout initialisation. It is important that it be applied to the ReLU layer that comes before the dropout layer.

```python
def dropout_(tensor, p=0, mode='fan_in', nonlinearity='dropout_relu'):
    r"""Fills the input `Tensor` with values according to the method
    described in "Critical initialisation for deep signal propagation in
noisy rectifier neural networks" - Pretorius, A. et al. (2018). The resulting tensor will have values sampled from
    :math:`\mathcal{N}(0, \text{std})` where

    .. math::
        \text{std} = \sqrt{\frac{2 \times (1-p)}{\text{fan\_in}}}

    Args:
        tensor: an n-dimensional `torch.Tensor`
        p: probability of an element to be zeroed. This should be set equal to the dropout rate p in the subsequent dropout layer. 
        mode: either 'fan_in' (default) or 'fan_out'. Choosing `fan_in`
            preserves the magnitude of the variance of the weights in the
            forward pass. Choosing `fan_out` preserves the magnitudes in the
            backwards pass.
        nonlinearity: the non-linear function (`nn.functional` name),
            recommended to use only with 'relu'.

    Examples:
        >>> w = torch.empty(3, 5)
        >>> nn.init.dropout_(w, p=0.6, mode='fan_in', nonlinearity='relu')
    """
    fan = _calculate_correct_fan(tensor, mode)
    gain = calculate_gain(nonlinearity, p)
    std = gain / math.sqrt(fan)
    with torch.no_grad():
        return tensor.normal_(0, std)
```

## Alternatives

Please feel free to suggest any alternative implementations. :)

**Comment by Kaixhin** [comment]
This could be good for https://github.com/pytorch/contrib, which takes "implementations of ideas from recent machine learning papers".

**Comment by soumith** [comment]
closing via Kai's comment
